Dev 1

kateraawi opened the pull request.

Commit:
Added gitignore

Commit:
Added dbconfig, created models

Commit:
Added new methods for models, refactoring

kateraawi committed:
Refactoring:
- now Task->getUsers and User->getTasks returns values not changing the object (expandSelf method still changing object)
Added field repeats and completed for Task model

kateraawi committed:
Added attach/detach methods for Task and User

osipov-dr commented on index.php:
```diff
@@ -1,2 +1,17 @@
+    ini_set('display_startup_errors', '1');
+    error_reporting(E_ALL);
+
+    require_once ('./services/dbconfig.php');
```
лучше использовать __DIR__, dirname 
почитай

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    function defineSelf(){
+        global $link;
```
точно не link
connection мб

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        $this->completed = $completed;
+    }
+
+    function defineSelf(){
```
странное название
я, по крайней мере, не понял, что она делает, пока внутрь не зашёл

osipov-dr replied:
{ со следующей строки

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        $this->repeats = $this->getRepeats();
+    }
+
+    function getJoined() {
```
get joined что?)

osipov-dr replied:
ок, класс называется Task, тогда я предполагаю, что заджойнены будут задачи, но с чем?
да и как из названия понять, что много тасков джойнятся, а не один

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        return mysqli_query($link, "SELECT * FROM tasks INNER JOIN users_tasks ON tasks.id = users_tasks.task_id AND tasks.id = $this->id INNER JOIN users ON users.id = users_tasks.user_id");
+    }
+
+    private function getUsers(){
```
функции описываются в порядке вызова сверху-вниз, то есть getUsers должна быть сразу после defineSelf, а потом getRepeats

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+            array_push($users, $user);
+        }
+
+        $this->users = $users;
```
не очень понял, что у тебя за структура и что делает эта функция
особенно после прочтения getUsers. и там, и там набивается массив users

kateraawi replied:
Здесь такая идея: у каждого Task есть поле users, а у каждого User есть поле tasks. Так вот, getUsers/Tasks записывают только массив id User/Task в это поле, но никак не объект, т. к. если бы мы сразу получали массив объектов User и Task, то при получении User бы стал подхватывать tasks, которые стали бы подхватывать users, которые... В общем, получилась бы рекурсия. expandSelf раскрывает как раз массив users/tasks, делая из этого объекты, у которых опять в tasks/users массив id

osipov-dr replied:
звучит интересно, а зачем это нужно? в том плане, что мб для получения задач по человеку просто получать их по его id запросом? или я опять чего-то не понял)

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+
+    function addSelf() {
+        global $link;
+        $descr = mysqli_real_escape_string($link, $this->descr);
```
не надо сокращений, кроме совсем распространённых, типа int от integer
а желательно вообще без них

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        $do_from = mysqli_real_escape_string($link, $this->do_from);
+        $do_to = mysqli_real_escape_string($link, $this->do_to);
+        $period_days = mysqli_real_escape_string($link, $this->period_days);
+        $period_qua = mysqli_real_escape_string($link, $this->period_qua);
```
аналогично. что такое qua знаешь только ты

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        $period_days = mysqli_real_escape_string($link, $this->period_days);
+        $period_qua = mysqli_real_escape_string($link, $this->period_qua);
+        $completed = mysqli_real_escape_string($link, $this->completed);
+        mysqli_query($link, "INSERT into tasks (descr, do_from, do_to, period_days, period_qua, completed) VALUES ( $descr, $do_from, $do_to, $period_days, $period_qua, $completed)");
```
используй подготовленные запросы

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+
+    function expandSelf(){
+        $users = [];
+        foreach ($this->users as $userID) {
```
не ID, а Id
только первая заглавная, даже у сокращений
чтобы отделить от следующего слова

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        mysqli_query($link, "INSERT into tasks (descr, do_from, do_to, period_days, period_qua, completed) VALUES ( $descr, $do_from, $do_to, $period_days, $period_qua, $completed)");
+    }
+
+    function getRepeats(){
```
вроде ничего, но хочется поменять название как-то
побольше раскрыть
понятно, что к таскам относится, но что такое repeats не раскрыто
у тебя это периоды?

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        return $repeats;
+    }
+
+    function attachUser($id){
```
id таска или юзера?
не должен я в метод заглядывать или задумываться лишний раз, чтобы это понять
attach не сложное слово, но я бы использовал add: не все хорошо знают английский, а add все знают

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        global $link;
+        $result = mysqli_query($link, "SELECT * FROM users_tasks WHERE user_id=$id AND task_id=$this->id");
+        $result = mysqli_fetch_assoc($result);
+        if ($result == null) {
```
== серьёзно?)

osipov-dr commented on models/Task.php:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    function detachUser($id){
```
слово сложное
мб unlink
опять вопрос про id

osipov-dr commented on models/User.php:
```diff
@@ -0,0 +1,69 @@
+<?php
+
+require_once (__DIR__.'/../services/dbconfig.php');
```
о, dir)
пробелы вокруг оператора конкатенации 
dirname можно для поднятия по структуре

osipov-dr commented on models/User.php:
```diff
@@ -0,0 +1,69 @@
+
+    private function getJoined(){
+        global $link;
+        return mysqli_query($link, "SELECT * FROM users INNER JOIN users_tasks ON users.id = users_tasks.user_id AND users.id = $this->id  INNER JOIN tasks ON tasks.id = users_tasks.task_id");
```
запросы не очень длинные, но я бы вынес в отдельные файлы

osipov-dr commented on models/User.php:
```diff
@@ -0,0 +1,69 @@
+
+    function addSelf() {
+        global $link;
+        mysqli_query($link, "INSERT into users (name) VALUES (" . mysqli_real_escape_string($link, $this->name) . ")");
```
выглядит стрёмно

kateraawi replied:
Использовать подготовленный запрос?

osipov-dr replied:
однозначно

osipov-dr commented on models/User.php:
```diff
@@ -0,0 +1,69 @@
+
+    function detachTask($id){
+        global $link;
+        mysqli_query($link, "DELETE FROM users_tasks WHERE user_id=$this->id AND task_id=$id");
```
а если ошибка при удалении?

kateraawi replied:
Транзакции или что-то другое?

osipov-dr replied:
они самые

kateraawi committed:
dbconfig.php:
- renamed link to connection

Task.php User.php:
- Deprecated expandSelf method for Task and User
- defineSelf was renamed to dbConstruct for Task and User
- attach/detach methods were renamed to add/unlink Task and User
- getRepeats was renamed to getRepeatedPeriods
- used transactions
- used prepared queries

index.php:
- minor changes due to previous changes

kateraawi committed:
Refactoring, new methods
